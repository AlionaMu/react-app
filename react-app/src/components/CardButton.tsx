const CardButton = (props: any) => {

  const openModal = () => {
    props.setModal(true);
    props.setCardId(props.cardId);
  }

  return (
    <>
      <button className='card__button button' onClick={openModal} >{props.text}</button>
    </>
  );
};
    
export default CardButton;