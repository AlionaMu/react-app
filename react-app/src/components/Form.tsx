import React from 'react';

export interface FormInfo {
  input: string;
  date: string;
  gender: string;
  img: string;
  checkbox: boolean;
  select: string;
}

function validate(input: string, gender: string, date: string, img: string, select: string) {
  const errors: string[] = [];

  if (!input) {
    errors.push('Name can\'t be empty');
  }

  if (!gender) {
    errors.push('Gender can\'t be empty');
  }
  
  if (!date) {
    errors.push('Date can\'t be empty');
  }

  if (!img) {
    errors.push('Upload file please');
  }

  if (select === 'Select country') {
    errors.push('Select country please');
  }

  return errors;
}

export class Form extends React.Component<Record<any, any>> {  
  public input: any;
  public date: any;
  public gender: any;
  public checkbox: any;
  public switchMale: any;
  public switchFemale: any;
  public img: any;
  public card: FormInfo;
  public select: any;
  public options: any[];
  
  constructor(props: any) {
    super(props);
    this.input = React.createRef();
    this.date = React.createRef();
    this.checkbox = React.createRef();
    this.switchMale = React.createRef();
    this.switchFemale = React.createRef();
    this.img = React.createRef();
  
    this.select = React.createRef();
    this.options = [
     {value: 'Belarus', title: 'Belarus'},
     {value: 'India', title: 'India'},
     {value: 'Australia', title: 'Australia'},
     {value: 'Venesuela', title: 'Venesuela'},
    ];

    this.state = {
      errors: [],
      disabled: true
    };
  
    this.handleSubmit = this.handleSubmit.bind(this);
    this.setButtonAble = this.setButtonAble.bind(this);
  
    this.card = {
       input: this.input,
       date: this.date,
       gender: this.gender,
       img: this.img,
       checkbox: this.checkbox,
       select: this.select
     };
  }

  setButtonAble() {
    this.setState({...this.state, disabled: false });
  }
  
  handleSubmit(event: any) {
    event.preventDefault();
    this.switchMale.current.checked === true ?
    this.gender = 'male' :
    this.switchFemale.current.checked ?
    this.gender = 'female' : ''

    this.card.gender = this.gender;
    this.card.input = this.input.current.value;
    this.card.date = this.date.current.value;
    this.card.img = URL.createObjectURL(this.img.current?.files?.[0])
    this.card.checkbox = this.checkbox.current.checked;
    this.card.select = this.select.current.value;
    
    const errors: string[] = validate(this.card.input, this.card.gender, this.card.date, this.card.img, this.card.select );
    if (errors.length > 0) {
      this.setState({...this.state, errors: errors }); 
      return;
    } else { this.props.create(event, this.card) }

  }

  render() {
    const { errors } = this.state as any;
    const { disabled } = this.state as any;

    return (
        <form className="form" onSubmit={this.handleSubmit} onChange={this.setButtonAble}>
            {errors.map((error: any) => (
          <p className="form__error" key={error}>Error: {error}</p>
        ))}
          <label>
            Name:
            <input type="text" ref={this.input} className="form__input" />
          </label>
          <label>
            Date:
            <input type="date" ref={this.date} className="form__input" />
          </label>
          <fieldset>
            <label className="form__switch">
              Male
              <input value="male" name="gender" type="radio" ref={this.switchMale} /> 
            </label>
            <label className="form__switch">
              Female
              <input value="female" name="gender" type="radio" ref={this.switchFemale} />
            </label>
          </fieldset>
          <select ref={this.select}>
            <option>Select country</option>
              {this.options.map((option: any) =>
              <option value={option.value} key={option.value}>
               {option.title}
              </option>
            )}
          </select>
          <p>Do you agree to the processing of data?</p>
          <label className="form__checkbox">
            <input type="checkbox" ref={this.checkbox} />
          </label>
          <label>
            Upload file:
            <input className='form__file-input' type="file" ref={this.img} />
          </label>
          <input type="submit" value="Submit" className='button form__button' disabled={disabled} />
        </form>
    );
  }
}
