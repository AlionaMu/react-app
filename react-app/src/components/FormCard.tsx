const FormCard = (props: any) => {
    return (
      <div className='card' id="1" key={props}>
       
          <div className='card__contain'>
            <div className='card__text'>
              <span className='card__text_title'>{props.props.input}</span>
              <span className='card__text_title'>{props.props.date}</span>
              <span className='card__text_title'>{props.props.select}</span>
              <span className='card__text_title'>{props.props.checkbox}</span>
              <span className='card__text_title'>{props.props.gender}</span>
              <img src={props.props.img} className='form-card__image' alt="avatar"></img>
   
            </div>
          </div>
      </div>
    );
  };
  
  export default FormCard;
  