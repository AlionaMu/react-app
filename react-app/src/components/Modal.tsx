import MoreInfoCard from './MoreInfoCard';

const Modal = ({visible, setVisible, cardId, cardInfo}: any) => {  

    const classes = ['modal'];

    if(visible) {
      classes.push('active');
    }
    return (
      <div className={classes.join(' ')} onClick={() => setVisible(false)}>
        <div className="modal__content" onClick={(e) => {e.stopPropagation()}}>
        <MoreInfoCard cardId={cardId} setModal={setVisible} cardInfo={cardInfo}></MoreInfoCard>
        </div>
      </div>
    );
  };
    
  export default Modal;
  