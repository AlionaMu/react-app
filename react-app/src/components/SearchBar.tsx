import FetchService from '../API/FetchService';

export function SearchBar(props: any) {

  const onKeyPressHandler = (event: any) => {
   
    if (event.which === 13) {
      props.setLoading(true);

// setTimeout is added to demonstrate Loading 

      setTimeout(() => {
       
        FetchService.getPosts(event.target.value).then((res) => {
          props.setCardsList([...res.items]);
          props.setLoading(false);
        })

      }, 3000)

    }
  }

  return (
    <div>
      <input 
        className="search-bar" 
        data-testid="simple-input" 
        type="text" 
        placeholder='search'
        onKeyPress={(event: any) => onKeyPressHandler(event)}
      >
      </input>
      <button className="search-bar__button button">SEARCH</button> 
    </div>
  )
}
