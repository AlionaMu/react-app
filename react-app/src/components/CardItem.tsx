import CardButton from './CardButton';

export interface Item {
  etag: string;
  id: ItemId;
  kind: string;
  snippet: ItemSnippet;
}

export interface ItemId {
  id: string;
  videoId: string;
}

export interface ItemSnippet {
  channelId?: string;
  channelTitle: string;
  description?: string;
  liveBroadcastContent?: string;
  publishTime?: string;
  publishedAt?: string;
  thumbnails?: Thumb;
  title: string
}
 
export interface Thumb {
  default: ThumbContent;
  medium: ThumbContent;
  high: ThumbContent;
}

export interface ThumbContent {
  height: number;
  url: string;
  width: number;
}

const CardItem = (props: any) => {
  return (
    <div className='card' id={props.props.id}>
        <img src={props.props.snippet.thumbnails.default.url} className='card__image' alt="fruit"></img>
        <div className='card__contain'>
          <span className='card__text_title'><b>Video Title: </b>{props.props.snippet.title}</span>
          <span className='card__text_species'><b>Channel Title: </b>{props.props.snippet.channelTitle}</span>
        </div>
        <CardButton
          modal={props.modal} 
          setModal={props.setModal} 
          text={'more info...'} 
          cardId={props.props.id} 
          setCardId={props.setCardId}>
        </CardButton>
    </div>
  );
};

export default CardItem;
  