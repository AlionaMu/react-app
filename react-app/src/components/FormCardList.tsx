import FormCard from './FormCard';

const FormCardList = (props: any) => {
    return (
              <div>
                { props.memoList ?
                  <div className="form-cards__container">
                    { props.memoList.map((item: any, i: number) => {
                    return <FormCard props={item} key={i + 1} src={item.src} />
                    })}
                  </div> :
                  <span> List is empty </span>
                }
              </div>
          );
  };
  
  export default FormCardList;
