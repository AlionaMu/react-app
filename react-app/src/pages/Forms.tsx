import { useMemo, useState } from 'react';
import { Form, FormInfo } from '../components/Form';
import FormCardList from '../components/FormCardList';

function Forms() {
  const arr: FormInfo[] = [];
  const [list, setList] = useState(arr);

  const memoList: FormInfo[] = useMemo(() => {
    return list;
  }, [list, setList]);

  const formProps = { 
    memoList: memoList,
    setList: setList
  }

  const creatFormCard = (event: Event, newCard: FormInfo) => {
    event.preventDefault();
    setList([...list, newCard]);
  }

  return (
    <div>
      <h1>FORMS</h1>
      <Form create={creatFormCard} />
      <FormCardList {...formProps} />
    </div>
  )
}

export default Forms;
