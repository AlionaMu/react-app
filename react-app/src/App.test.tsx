import { fireEvent, render, screen } from '@testing-library/react';
import { SearchBar } from './components/SearchBar';
import Forms from './pages/Forms';
import CardList from './components/CardList';

describe('SearchBar', () => {
  it('renders SearchBar component', () => {
    render(<SearchBar />);
    expect(screen.getByText(/search/i)).toBeInTheDocument();
    screen.debug();
    expect(screen.getByPlaceholderText(/s/i)).toBeInTheDocument();
    screen.debug();
    expect(screen.getByPlaceholderText(/search/i)).toHaveClass('search-bar')
  })
})

describe('events', () => {
  it('click', () => {
    const handleChange = jest.fn();
    const { container } = render(<input type='checkbox' onChange={handleChange} />)
    const checkbox = container.firstChild as ChildNode;
    expect(checkbox).not.toBeChecked();
    fireEvent.click(checkbox);
    expect(handleChange).toHaveBeenCalledTimes(1);
    expect(checkbox).toBeChecked();
  });

  it('input focus', () => {
    const { getByTestId } = render(<SearchBar />);
    const input = getByTestId('simple-input');
    expect(input).not.toHaveFocus();
    input.focus();
    expect(input).toHaveFocus();
  })
})

describe('Forms', () => {
  it('renders Form component', () => {
    render(<Forms />);
    expect(screen.getByText(/Name/i)).toBeInTheDocument();
    screen.debug();
  });
}); 

const mockedFetch = jest.fn(() =>
  Promise.resolve({
    json: () => Promise.resolve({}),
  })
) as jest.Mock;

global.fetch = mockedFetch;

test('Renders the component with posts', async () => {
  render(<CardList />);
  expect(fetch).toHaveBeenCalledTimes(0);
});


 
