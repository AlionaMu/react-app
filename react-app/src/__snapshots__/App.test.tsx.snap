// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders learn react link 1`] = `
<DocumentFragment>
  <div>
    <input
      class="search-bar"
      placeholder="search"
      type="text"
    />
    <button
      class="search-bar__button button"
    >
      search
    </button>
  </div>
</DocumentFragment>
`;
